### PR TITLE
Peer Manager

### PR DESCRIPTION
**Pull request by oskin1**
#2tym827

- [x] PeerDB
- [x] Peers State Machine
- [x] PeerManager API
- [x] PeerManager Behavior
- [x] PeerManager Stream

### CODE REVIEW

File: spectrum-network/src/peer/peer_state.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+impl<'a> NotConnectedPeer<'a> {
+    fn to_peer_id(self) -> PeerId {
+        self.peer_id.into_owned()
+    }
+}
+
+pub enum PeerInState<'a> {
+    /// We are connected to this peer.
+    Connected(ConnectedPeer<'a>),
+    /// We are connected to this peer.
```
Comment (techraed) [inline spectrum-network/src/peer/peer_state.rs]: typo?

File: spectrum-network/src/peer/peer_state.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
+use std::borrow::{Borrow, Cow};
+use libp2p::PeerId;
+use crate::peer::data::{PeerConnState, PeerInfo};
+use crate::peer::peer_store::{PeerStore, PeerStoreRejection};
+use crate::peer::types::Reputation;
```
Comment (techraed) [inline spectrum-network/src/peer/peer_state.rs]: can be united like
```suggestion
use crate::peer::{
    data::{PeerConnState, PeerInfo},
    peer_store::{PeerStore, PeerStoreRejection},
    types::Reputation,
};
```

File: spectrum-network/src/peer/peer_store.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+impl PeerStore for InMemoryPeerStore {
+    fn add(&mut self, peer_id: &PeerId, peer: PeerInfo) -> Result<&mut PeerInfo, PeerStoreRejection> {
+        if self.peers.len() < self.conf.capacity {
+            if self.peers.contains_key(&peer_id) {
+                Err(PeerStoreRejection::AlreadyExists)
+            } else {
+                self.peers.insert(peer_id.clone(), peer);
+                Ok(self.peers.get_mut(&peer_id).unwrap())
```
Comment (techraed) [inline spectrum-network/src/peer/peer_store.rs]:
```suggestion
                Ok(self.peers.entry(peer_id.clone()).or_insert(peer))
```

File: spectrum-network/src/peer/peer_store.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    capacity: usize,
+}
+
+pub struct InMemoryPeerStore {
+    peers: HashMap<PeerId, PeerInfo>,
+    conf: PeerStoreConfig,
+}
+
+impl PeerStore for InMemoryPeerStore {
+    fn add(&mut self, peer_id: &PeerId, peer: PeerInfo) -> Result<&mut PeerInfo, PeerStoreRejection> {
```
Comment (techraed) [inline spectrum-network/src/peer/peer_store.rs]: `PeerId` is a `Copy` type. I would  rather remove the reference, or use `*peer_id` when needed

File: spectrum-network/src/peer/peer_state.rs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+pub struct ConnectedPeer<'a> {
+    peer_id: Cow<'a, PeerId>,
```
Comment (techraed) [inline spectrum-network/src/peer/peer_state.rs]: `PeerId` is a copy type. That fact can bring lots of simplicity to the code by removing the lifetimes and `Cow`.

File: spectrum-network/src/peer/peer_state.rs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    fn add_peer(&mut self, peer_id: PeerId, is_reserved: bool) -> Result<NotConnectedPeer, PeerStoreRejection> {
+        self.store
+            .add(&peer_id, PeerInfo::new(is_reserved))
+            .map(|p| NotConnectedPeer { peer_id: Cow::Owned(peer_id), peer_info: p })
+    }
+
+    fn connect_to_peer<'a>(&mut self, peer: NotConnectedPeer<'a>) -> ConnectedPeer<'a> {
+        peer.peer_info.num_connections += 1;
```
Comment (techraed) [inline spectrum-network/src/peer/peer_state.rs]: consider using `saturating_add` in the future

File: spectrum-network/src/peer/peer_state.rs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        peer.peer_info.state = PeerConnState::Connected;
+        ConnectedPeer { peer_id: peer.peer_id, peer_info: peer.peer_info }
+    }
+
+    fn disconnect_peer<'a>(&mut self, peer: ConnectedPeer<'a>) -> NotConnectedPeer<'a> {
```
Comment (techraed) [inline spectrum-network/src/peer/peer_state.rs]: shall we decrease num of connections?

File: spectrum-network/src/peer/peer_state.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+        peer.peer_info.state = PeerConnState::NotConnected;
+        NotConnectedPeer { peer_id: peer.peer_id, peer_info: peer.peer_info }
+    }
+
+    fn forget_peer(&mut self, peer: NotConnectedPeer) {
+        self.store.drop(peer.peer_id.borrow());
+        ()
```
Comment (techraed) [inline spectrum-network/src/peer/peer_state.rs]: not needed

File: spectrum-network/src/peer/peer_state.rs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        ConnectedPeer { peer_id: peer.peer_id, peer_info: peer.peer_info }
+    }
+
+    fn disconnect_peer<'a>(&mut self, peer: ConnectedPeer<'a>) -> NotConnectedPeer<'a> {
+        peer.peer_info.state = PeerConnState::NotConnected;
+        NotConnectedPeer { peer_id: peer.peer_id, peer_info: peer.peer_info }
+    }
+
+    fn forget_peer(&mut self, peer: NotConnectedPeer) {
+        self.store.drop(peer.peer_id.borrow());
```
Comment (techraed) [inline spectrum-network/src/peer/peer_state.rs]:
```suggestion
        self.store.drop(&peer.peer_id);
```

File: spectrum-network/src/peer/types.rs [hunk 10]
```diff
@@ -0,0 +1,26 @@
+/// Opaque identifier for an incoming connection. Allocated by the network.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct IncomingIndex(pub u64);
```
Comment (techraed) [inline spectrum-network/src/peer/types.rs]: Better to hid internal type. If it's needed to return it, then better follow `into_inner(self)` "convention"

**Review by techraed — COMMENTED** [review]
please use `cargo fmt`

File: spectrum-network/src/peer/peer_state.rs (hunk 1 above)
Comment (oskin1) [reply spectrum-network/src/peer/peer_state.rs]: Yep

File: spectrum-network/src/peer/peer_state.rs (hunk 7 above)
Comment (oskin1) [reply spectrum-network/src/peer/peer_state.rs]: Counter of connections to the peer is monotonically increasing

**Comment by yasha-black** [comment]
Task linked: [CU-2tym827 Implement Peer Manager](https://app.clickup.com/t/2tym827)

**Pull request merged**